// Data.String.CodePoints/index.js

// | These functions allow PureScript strings to be treated as if they were

// | sequences of Unicode code points instead of their true underlying

// | implementation (sequences of UTF-16 code units). For nearly all uses of

// | strings, these functions should be preferred over the ones in Data.String.
"use strict";
var $foreign = require("./foreign");
var Control_Semigroupoid = require("../Control.Semigroupoid");
var Data_Array = require("../Data.Array");
var Data_Char = require("../Data.Char");
var Data_Eq = require("../Data.Eq");
var Data_EuclideanRing = require("../Data.EuclideanRing");
var Data_Functor = require("../Data.Functor");
var Data_HeytingAlgebra = require("../Data.HeytingAlgebra");
var Data_Maybe = require("../Data.Maybe");
var Data_Ord = require("../Data.Ord");
var Data_Ring = require("../Data.Ring");
var Data_Semigroup = require("../Data.Semigroup");
var Data_Semiring = require("../Data.Semiring");
var Data_String = require("../Data.String");
var Data_String_Unsafe = require("../Data.String.Unsafe");
var Data_Tuple = require("../Data.Tuple");
var Data_Unfoldable = require("../Data.Unfoldable");
var Prelude = require("../Prelude");

// | CodePoint is an Int bounded between 0 and 0x10FFFF, corresponding to

// | Unicode code points.
var CodePoint = function (x) {
    return x;
};
var unsurrogate = function (lead) {
    return function (trail) {
        return (((lead - 55296 | 0) * 1024 | 0) + (trail - 56320 | 0) | 0) + 65536 | 0;
    };
};
var isTrail = function (cu) {
    return 56320 <= cu && cu <= 57343;
};
var isLead = function (cu) {
    return 55296 <= cu && cu <= 56319;
};

// | Returns a record with the first code point and the remaining code points

// | of the string. Returns Nothing if the string is empty. Operates in

// | constant space and time.
var uncons = function (s) {
    var v = Data_String.length(s);
    if (v === 0) {
        return Data_Maybe.Nothing.value;
    };
    if (v === 1) {
        return new Data_Maybe.Just({
            head: Data_String_Unsafe.charCodeAt(0)(s), 
            tail: ""
        });
    };
    var cu0 = Data_String_Unsafe.charCodeAt(0)(s);
    var cu1 = Data_String_Unsafe.charCodeAt(1)(s);
    var $19 = isLead(cu0) && isTrail(cu1);
    if ($19) {
        return new Data_Maybe.Just({
            head: unsurrogate(cu0)(cu1), 
            tail: Data_String.drop(2)(s)
        });
    };
    return new Data_Maybe.Just({
        head: cu0, 
        tail: Data_String.drop(1)(s)
    });
};
var unconsButWithTuple = function (s) {
    return Data_Functor.map(Data_Maybe.functorMaybe)(function (v) {
        return new Data_Tuple.Tuple(v.head, v.tail);
    })(uncons(s));
};
var toCodePointArrayFallback = function (s) {
    return Data_Unfoldable.unfoldr(Data_Unfoldable.unfoldableArray)(unconsButWithTuple)(s);
};
var unsafeCodePointAt0Fallback = function (s) {
    var cu0 = Data_String_Unsafe.charCodeAt(0)(s);
    var cu1 = Data_String_Unsafe.charCodeAt(1)(s);
    var $23 = isLead(cu0) && isTrail(cu1);
    if ($23) {
        return unsurrogate(cu0)(cu1);
    };
    return cu0;
};

// WARN: this function expects the String parameter to be non-empty
var unsafeCodePointAt0 = $foreign._unsafeCodePointAt0(unsafeCodePointAt0Fallback);

// | Creates an array of code points from a string. Operates in space and time

// | linear to the length of the string.
var toCodePointArray = $foreign._toCodePointArray(toCodePointArrayFallback)(unsafeCodePointAt0);

// | Returns the number of code points in the string. Operates in constant

// | space and in time linear to the length of the string.
var length = function ($50) {
    return Data_Array.length(toCodePointArray($50));
};

// | Returns the number of code points preceding the last match of the given

// | pattern in the string. Returns Nothing when no matches are found.
var lastIndexOf = function (p) {
    return function (s) {
        return Data_Functor.map(Data_Maybe.functorMaybe)(function (i) {
            return length(Data_String.take(i)(s));
        })(Data_String.lastIndexOf(p)(s));
    };
};

// | Returns the number of code points preceding the first match of the given

// | pattern in the string. Returns Nothing when no matches are found.
var indexOf = function (p) {
    return function (s) {
        return Data_Functor.map(Data_Maybe.functorMaybe)(function (i) {
            return length(Data_String.take(i)(s));
        })(Data_String.indexOf(p)(s));
    };
};
var fromCharCode = function ($51) {
    return Data_String.singleton(Data_Char.fromCharCode($51));
};
var singletonFallback = function (v) {
    if (v <= 65535) {
        return fromCharCode(v);
    };
    var lead = ((v - 65536 | 0) / 1024 | 0) + 55296 | 0;
    var trail = (v - 65536 | 0) % 1024 + 56320 | 0;
    return fromCharCode(lead) + fromCharCode(trail);
};

// | Creates a string from an array of code points. Operates in space and time

// | linear to the length of the array.
var fromCodePointArray = $foreign._fromCodePointArray(singletonFallback);

// | Returns a record with strings created from the code points on either side

// | of the given index. If the index is not within the string, Nothing is

// | returned.
var splitAt = function (i) {
    return function (s) {
        var cps = toCodePointArray(s);
        var $25 = i < 0 || Data_Array.length(cps) < i;
        if ($25) {
            return Data_Maybe.Nothing.value;
        };
        return new Data_Maybe.Just({
            before: fromCodePointArray(Data_Array.take(i)(cps)), 
            after: fromCodePointArray(Data_Array.drop(i)(cps))
        });
    };
};

// | Creates a string containing just the given code point. Operates in

// | constant space and time.
var singleton = $foreign._singleton(singletonFallback);
var takeFallback = function (n) {
    return function (v) {
        if (n < 1) {
            return "";
        };
        var v1 = uncons(v);
        if (v1 instanceof Data_Maybe.Just) {
            return singleton(v1.value0.head) + takeFallback(n - 1 | 0)(v1.value0.tail);
        };
        return v;
    };
};

// | Returns a string containing the given number of code points from the

// | beginning of the given string. If the string does not have that many code

// | points, returns the empty string. Operates in constant space and in time

// | linear to the given number.
var take = $foreign._take(takeFallback);

// | Returns the number of code points preceding the first match of the given

// | pattern in the string. Pattern matches following the given index will be

// | ignored. Returns Nothing when no matches are found.
var lastIndexOf$prime = function (p) {
    return function (i) {
        return function (s) {
            var i$prime = Data_String.length(take(i)(s));
            return Data_Functor.map(Data_Maybe.functorMaybe)(function (k) {
                return length(Data_String.take(k)(s));
            })(Data_String["lastIndexOf'"](p)(i$prime)(s));
        };
    };
};
var eqCodePoint = new Data_Eq.Eq(function (x) {
    return function (y) {
        return x === y;
    };
});
var ordCodePoint = new Data_Ord.Ord(function () {
    return eqCodePoint;
}, function (x) {
    return function (y) {
        return Data_Ord.compare(Data_Ord.ordInt)(x)(y);
    };
});

// | Drops the given number of code points from the beginning of the string. If

// | the string does not have that many code points, returns the empty string.

// | Operates in constant space and in time linear to the given number.
var drop = function (n) {
    return function (s) {
        return Data_String.drop(Data_String.length(take(n)(s)))(s);
    };
};

// | Returns the number of code points preceding the first match of the given

// | pattern in the string. Pattern matches preceding the given index will be

// | ignored. Returns Nothing when no matches are found.
var indexOf$prime = function (p) {
    return function (i) {
        return function (s) {
            var s$prime = drop(i)(s);
            return Data_Functor.map(Data_Maybe.functorMaybe)(function (k) {
                return i + length(Data_String.take(k)(s$prime)) | 0;
            })(Data_String.indexOf(p)(s$prime));
        };
    };
};
var countTail = function ($copy_p) {
    return function ($copy_s) {
        return function ($copy_accum) {
            var $tco_var_p = $copy_p;
            var $tco_var_s = $copy_s;
            var $tco_done = false;
            var $tco_result;
            function $tco_loop(p, s, accum) {
                var v = uncons(s);
                if (v instanceof Data_Maybe.Just) {
                    var $37 = p(v.value0.head);
                    if ($37) {
                        $tco_var_p = p;
                        $tco_var_s = v.value0.tail;
                        $copy_accum = accum + 1 | 0;
                        return;
                    };
                    $tco_done = true;
                    return accum;
                };
                $tco_done = true;
                return accum;
            };
            while (!$tco_done) {
                $tco_result = $tco_loop($tco_var_p, $tco_var_s, $copy_accum);
            };
            return $tco_result;
        };
    };
};
var countFallback = function (p) {
    return function (s) {
        return countTail(p)(s)(0);
    };
};

// | Returns the number of code points in the leading sequence of code points

// | which all match the given predicate. Operates in constant space and in

// | time linear to the length of the string.
var count = $foreign._count(countFallback)(unsafeCodePointAt0);

// | Drops the leading sequence of code points which all match the given

// | predicate from the string. Operates in constant space and in time linear

// | to the length of the string.
var dropWhile = function (p) {
    return function (s) {
        return drop(count(p)(s))(s);
    };
};

// | Returns a string containing the leading sequence of code points which all

// | match the given predicate from the string. Operates in constant space and

// | in time linear to the length of the string.
var takeWhile = function (p) {
    return function (s) {
        return take(count(p)(s))(s);
    };
};
var codePointToInt = function (v) {
    return v;
};

// I would prefer that this smart constructor not need to exist and instead

// CodePoint just implements Enum, but the Enum module already depends on this

// one. To avoid the circular dependency, we just expose these two functions.
var codePointFromInt = function (n) {
    if (0 <= n && n <= 1114111) {
        return new Data_Maybe.Just(n);
    };
    return Data_Maybe.Nothing.value;
};
var codePointAtFallback = function ($copy_n) {
    return function ($copy_s) {
        var $tco_var_n = $copy_n;
        var $tco_done = false;
        var $tco_result;
        function $tco_loop(n, s) {
            var v = uncons(s);
            if (v instanceof Data_Maybe.Just) {
                var $44 = n === 0;
                if ($44) {
                    $tco_done = true;
                    return new Data_Maybe.Just(v.value0.head);
                };
                $tco_var_n = n - 1 | 0;
                $copy_s = v.value0.tail;
                return;
            };
            $tco_done = true;
            return Data_Maybe.Nothing.value;
        };
        while (!$tco_done) {
            $tco_result = $tco_loop($tco_var_n, $copy_s);
        };
        return $tco_result;
    };
};

// | Returns the first code point of the string after dropping the given number

// | of code points from the beginning, if there is such a code point. Operates

// | in constant space and in time linear to the given index.
var codePointAt = function (v) {
    return function (v1) {
        if (v < 0) {
            return Data_Maybe.Nothing.value;
        };
        if (v === 0 && v1 === "") {
            return Data_Maybe.Nothing.value;
        };
        if (v === 0) {
            return new Data_Maybe.Just(unsafeCodePointAt0(v1));
        };
        return $foreign._codePointAt(codePointAtFallback)(Data_Maybe.Just.create)(Data_Maybe.Nothing.value)(unsafeCodePointAt0)(v)(v1);
    };
};
module.exports = {
    codePointAt: codePointAt, 
    codePointFromInt: codePointFromInt, 
    codePointToInt: codePointToInt, 
    count: count, 
    drop: drop, 
    dropWhile: dropWhile, 
    fromCodePointArray: fromCodePointArray, 
    indexOf: indexOf, 
    "indexOf'": indexOf$prime, 
    lastIndexOf: lastIndexOf, 
    "lastIndexOf'": lastIndexOf$prime, 
    length: length, 
    singleton: singleton, 
    splitAt: splitAt, 
    take: take, 
    takeWhile: takeWhile, 
    toCodePointArray: toCodePointArray, 
    uncons: uncons, 
    eqCodePoint: eqCodePoint, 
    ordCodePoint: ordCodePoint
};
